common/models/wordbreakers/src/dict/index.ts

```diff
@@ -0,0 +1,362 @@
+
+// const DEFAULT_PARAMS = {
+
+// }
+
+export function splitOnWhitespace(text: string): Span[] {
+  const sections: Span[] = [];
+
+  let start = 0;
+
+  // Surrogate pairs will never overlap \u0020, so we don't need to be
+  // surrogate-pair aware here.
+  for(let index = 0; index < text.length + 1; index++) {
+    let char = (index == text.length ? ' ' : text.charAt(index));
+    if(char.match(/\s/) || char == '\u200c') {
+      if(start !== undefined) {
+        sections.push(new LazySpan(text, start, index));
+        start = undefined; // we do not emit whitespace tokens here.
+      }
+    } else if(start === undefined) {
+      start = index;
+    }
+  }
+
+  return sections;
+}
+
+/**
+ * Splits a string into its constituent codepoints.
+ *
+ * BMP chars will be placed within single-char strings; non-BMP chars will
+ * be represented by a single string with their surrogate-chars paired.
+ * @param text
+ * @returns
+ */
+export function splitOnCodepoints(text: string): string[] {
+  const splayed: string[] = [];
+  let tailIndex = 0;
+  let end = text.length;
+  while(tailIndex != end) {
+    const startIndex = tailIndex;
+    tailIndex = positionAfter(text, tailIndex);
+    splayed.push(text.substring(startIndex, tailIndex));
+  }
+
+  return splayed;
+}
+
+export type DictBreakerPath = {
+  /**
+   * The index of the character immediately before the most recently-available word boundary.
+   * Is set to -1 if no such boundary exists.
+   */
+  boundaryIndex: number;
+
+  // Could add a 'reference' if we create objects for each char in the context - such as for
+  // caching & reusing boundary info with future inputs.
+
+  /**
+   * An active traversal representing potential words that may become completed, starting
+   * immediately after the boundary indicated by `boundaryIndex`.
+   */
+  traversal: LexiconTraversal;
+
+  /**
+   * cost:  measured in -log(p) of each decision.
+   */
+  cost: number;
+
+  /**
+   * Indicates if this path's most recent traversal enforces a boundary without matching
+   * a word in the lexicon.
+   */
+  wasUnmatchedChar?: boolean;
+
+  /**
+   * The path object used to reach the previous boundary.
+   */
+  parent?: DictBreakerPath;
+}
+
+/**
+ * Provides dictionary-based wordbreaking assuming a LexiconTraversal can be specified for
+ * the dictionary.
+ * @param fullText The full context to be tokenized.
+ * @param dictRoot A LexiconTraversal interface from the active LexicalModel,
+ * allowing efficient dictionary lookups of encountered words.
+ * @returns
+ */
+export default function dict(fullText: string, dictRoot: LexiconTraversal): Span[] {
+  if(!dictRoot) {
+    throw new Error("Cannot use dictionary-based wordbreaker without `LexiconTraversal` dictionary access");
+  }
+
+  // Whenever we have a space or a ZWNJ (U+200C), we'll assume a 100%-confirmed wordbreak
+  // at that location.  We only need to "guess" at anything between 'em.
+  const sections = splitOnWhitespace(fullText);
```

Comment:
does this match @mhosken's "tent" model?

Reply:
... I'm not sure what you mean by "tent" model.  Could you elaborate?

Reply:
my recollection of the khmer word break improvements, some logic for detecting whether text was pre-broken.  and it affected more than just text surrounding the space (hence a 'tent').  Might need to ask him